charm/main.py

```diff
@@ -0,0 +1,218 @@
+# Copyright 2020 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charm
+
+
+import argparse
+import logging
+import os
+import sys
+
+from charm import logsetup
+
+logger = logging.getLogger(__name__)
+
+
+class CommandError(Exception):
+    """Base exception for all error commands.
+
+    It optionally receives an `retcode` parameter that will be the returned code
+    by the process on exit.
+    """
+    def __init__(self, message, retcode=-1):
+        self.retcode = retcode
+        super().__init__(message)
+
+
+class BaseCommand:
```

Comment:
is it worth making this an `abc.ABC` and tagging `run` with an `@abstractmethod`?

Reply:
Don't know. As only internal code would be inheriting from this, it pobably would not be necessary (if we were having external plugins, surely it would be different).